ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -155,6 +165,123 @@ static os_log_type_t KextLogLevelAsOSLogType(KextLog_Level level)
     }
 }
 
+static void LogPanics()
+{
+    DIR *dr = opendir(PanicLogDirectory);
+    if (dr == NULL)
+    {
+        ostringstream errorMessage;
+        errorMessage << "Unable to open " << PanicLogDirectory << ", errno=" << errno << ", errorstr=" << strerror(errno);
+        LogDaemonError(errorMessage.str());
+    }
+
+    // Look up the last run time
+    long lastRunTimestamp = 0;
+    if (DoesPathExist(PanicLogTimestampFile))
+    {
+        lastRunTimestamp = GetFileModifiedTime(PanicLogTimestampFile);
+    }
+
+    struct dirent *dirEntry;
+    while ((dirEntry = readdir(dr)) != NULL)
+    {
+        // Only check files
+        if(dirEntry->d_type == DT_REG)
+        {
+           // Check for .panic extension
+           const char *ext = strrchr(dirEntry->d_name,'.');
+           if(ext && strcmp(ext, ".panic") == 0)
+           {
+               // Get the full name of the panic file
+               string fullFileName = PanicLogDirectory;
+               fullFileName += "/";
+               fullFileName += dirEntry->d_name;
+
+               // Verify it's after the last timestamp
+               long panicLogTime = GetFileModifiedTime(fullFileName.c_str());
+               if (panicLogTime == 0)
+               {
+                   // Record error if we can't determine timestamp
+                   ostringstream errorMessage;
+                   errorMessage << "Unable to run stat on panic at " << fullFileName.c_str() << "\n";
+                   const char *errorMessageString = errorMessage.str().c_str();
+                   LogKextMessage(OS_LOG_TYPE_DEFAULT, 0, errorMessageString, static_cast<int>(strlen(errorMessageString)));
+                   continue;
+               }
+
+               if (panicLogTime >= lastRunTimestamp)
+               {
+                  // Verify 'org.vfsforgit.PrjFSKext(' occurs in the log
+                  FILE *fp = fopen(fullFileName.c_str(), "r");
+                  if (fp)
+                  {
+                      char* line = NULL;
+                      size_t len = 0;
+                      ssize_t read;
+                      while ((read = getline(&line, &len, fp)) != -1)
+                      {
+                          if (strstr(line, "org.vfsforgit.PrjFSKext(") != NULL)
+                          {
+                              ostringstream errorMessage;
+                              errorMessage << "Found panic at " << fullFileName.c_str() << "\n";
+                              LogDaemonError(errorMessage.str());
+                          }
```

Comment:
I'm not sure if it's possible for the string `org.vfsforgit.PrjFSKext(` to appear multiple times in a file; if it does, this will generate multiple log entries. We could `break;` out of the loop as soon as we fine at least one occurrence.

Reply:
This is a very good point!  Thanks @pmj!  As well as thank you for the nits on style.  I'll definitely fix!